asr-python-adapter-finetune-am-conformer-nemo.ipynb

```diff
@@ -0,0 +1,186 @@
+    "Adapters can be used with limited amounts of training data and compute budget, but they impose the restriction that the model's original vocabulary must be used - therefore a new characterset vocabulary / tokenizer cannot be used.\n",
+    "\n",
+    "Please refer to the **Advantages of Adapters** and **Limitations of Adapters** sections in the [NeMo Tutorial](#asr-with-adapters) for further details."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Preparing the Acoustic Encoder for Adapter Training\n",
+    "\n",
+    "Pre-trained models do not automatically support Adapter modules, so we must prepare the acoustic encoder prior to creating the model.\n",
+    "The steps for this procedure are detailed in the **Prepare the \"base\" model** section of the [NeMo Tutorial](#asr-with-adapters)."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Preparing the model and dataset for adaptation\n",
+    "\n",
+    "There are very few differences between fine-tuning and adapter training with respect to the preparation of the model, data loaders or optimization method.\n",
+    "\n",
+    "One significant difference is that there is no need to change the vocabulary of the pre-trained network, therefore there is no need to construct a new tokenizer or update the character vocabulary. If your customization dataset contains tokens that requires such changes (for example tokens that dont exist in the model), then you will need to either preprocess the text to remove such tokens, or instead perform fine-tuning (which may require substantially more data).\n",
+    "\n",
+    "Another minor difference is the use of spectrogram augmentation (SpecAugment). During fine-tuning, the amount of data is very little compared to when the model was pre-trained with thousands of hours of speech, so SpecAugment may have been used to prevent overfitting. For adapters, we recommend first disabling spec augmentation and trying to see if results are good enough, then adding it back slowly as needed to see if any improvements can be obtained.\n",
+    "\n",
+    "Please refer to the **Setup training and evaluation of the model** section, and particularly the **Setup Spectrogram Augmentation** section of the [NeMo Tutorial](#asr-with-adapters) for further details."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Creating and Training an Adapter\n",
+    "\n",
+    "The **Adapters: Creation and Preparation** section of the [NeMo Tutorial](#asr-with-adapters) details how to create an config object for a **LinearAdapter**, add it to the pretrained model, then freeze the original parameters of the model to train just the new parameters.\n",
+    "\n",
+    "All of this is a few lines of code, which will initialize your model so that it is prepared to train the adapter on new speech data."
```

Comment:
```suggestion
    "All of this is a few lines of code, which will initialize your model, so that it is prepared to train the adapter on new speech data."
```